File: scripts/DateTimePublisher.py
#!/usr/bin/python
import rospy
from openhab_msgs.msg import DateTimeCommand

class DateTimePublisher(object):
    """Node example class."""

    def __init__(self, item_name):

        self.item_name = item_name
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", DateTimeCommand, queue_size=10)
        self.rate = rospy.Rate(10) # 10hz

        # Initialize message variables.
        self.enable = False
        self.message = ""

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        """Turn on publisher."""
        self.enable = True
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", DateTimeCommand, queue_size=10)

        while not rospy.is_shutdown():

            self.message = DateTimeCommand()

            self.message.command = rospy.Time.now()

            self.message.header.stamp = rospy.Time.now()
            self.message.header.frame_id = "/base_link"
            self.message.item = self.item_name

            message = "Publishing to %s at %s: command = %s" % (self.message.item, rospy.get_time(), self.message.command)
            rospy.loginfo(message)

            self.pub.publish(self.message)
            self.rate.sleep()

    def stop(self):
        """Turn off publisher."""
        self.enable = False
        self.pub.unregister()

# Main function.
if __name__ == "__main__":
    # Initialize the node and name it.
    rospy.init_node("DateTimePublisherNode", anonymous=True)
    # Go to class functions that do all the heavy lifting.

    dateTimePublisher = DateTimePublisher("testDateTime")

    try:
        dateTimePublisher.start()
    except rospy.ROSInterruptException:
        pass
    # Allow ROS to go to all callbacks.
    # spin() simply keeps python from exiting until this node is stopped
    rospy.spin()

File: src/DateTimePublisher.cpp
#include <ros/ros.h>
#include <openhab_msgs/DateTimeCommand.h>

class DateTimePublisher
{
public:
    DateTimePublisher(const std::string& item_name)
        : item_name_(item_name),
          pub_(nh_.advertise<openhab_msgs::DateTimeCommand>("/openhab/items/" + item_name_ + "/command", 10)),
          rate_(10) // 10hz
    {
        if (enable_)
        {
            start();
        }
        else
        {
            stop();
        }
    }

    void start()
    {
        // Turn on publisher.
        enable_ = true;
        pub_ = nh_.advertise<openhab_msgs::DateTimeCommand>("/openhab/items/" + item_name_ + "/command", 10);

        while (ros::ok())
        {
            openhab_msgs::DateTimeCommand message;

            message.command = ros::Time::now();

            message.isnull = false;

            message.header.stamp = ros::Time::now();
            message.header.frame_id = "/base_link";
            message.item = item_name_;

            std::string message_str = "Publishing to " + message.item + " at " + std::to_string(ros::Time::now().toSec()) + ": command = " + std::to_string(message.command.toSec());
            ROS_INFO("%s", message_str.c_str());

            pub_.publish(message);
            rate_.sleep();
        }
    }

    void stop()
    {
        // Turn off publisher.
        enable_ = false;
        pub_.shutdown();
    }

private:
    std::string item_name_;
    ros::NodeHandle nh_;
    ros::Publisher pub_;
    ros::Rate rate_;
    bool enable_{false};
};

int main(int argc, char** argv)
{
    // Initialize the node and name it.
    ros::init(argc, argv, "DateTimePublisherNode", ros::init_options::AnonymousName);
    // Go to class functions that do all the heavy lifting.

    DateTimePublisher dateTimePublisher("testDateTime");

    try
    {
        dateTimePublisher.start();
    }
    catch (const ros::Exception& e)
    {
        ROS_ERROR("%s", e.what());
    }

    // Allow ROS to go to all callbacks.
    // spin() simply keeps the node from exiting until this node is stopped
    ros::spin();
    return 0;
}
